### DISCUSSION TITLE
운영체제 - 프로세스 - 호현&일구&성은

### DISCUSSION THREAD

**Discussion by seongeun42** [Shortudy-10th/tech-for-developer#108] [질의 응답]
* 프로세스간 통신이란 무엇이고 이를 위한 주요 방법이 무엇이 있는지 설명해주세요.
* 컨텍스트 스위칭에 대해 설명하고, 프로세스 간에는 어떻게 진행되는지 설명해주세요.
* 프로세스의 구조, 프로세스와 프로그램의 차이점을 말해주세요.
* 프로세스 종류는 무엇이 있는지 설명해주세요.
* 프로세스의 생성 과정에 대해 설명해주세요.

**Comment by seongeun42** [comment]
### ❓️ 프로세스간 통신이란 무엇이고 이를 위한 주요 방법이 무엇이 있는지 설명해주세요.

<br>

### 💬 답변
<blockquote>

* IPC(Inter Process Communication) 별 사용 설명
* PIPE 방식
   * 정의
      * 익명의 PIPE를 통해서 동일한 PPID를 가진 프로세스들 간에 단방향 통신을 지원
   * 구조
      *  FIFO 구조
      * 생성된 PIPE에 대하여 Write 또는 Read만 가능
   * 사용 시기
      * 부모 자식 프로세스간 통신 할 때 사용
   * 유의사항
      * 쌍방 통신을 위해서는 Write용 PIPE하나 Read PIPE 하나씩 만들어야 한다.
      * read()와 write()가 기본적으로 block 모드로 작동하기 때문으로 프로세스가 read 대기중이라면 read가 끝나기 전에는 write를 할수가 없게 된다.

![image](https://github.com/Shortudy-10th/tech-for-developer/assets/32262904/12bd0bdb-ca3f-40b6-a76c-417d91aec3e3)

* Named PIPE
   * 정의
      * 이름을 가진 PIPE를 통해서 프로세스들 간에 양 방향 통신을 지원
      * 기본적으로는 단 반향을 많이 사용합니다. 단 , 양 방향 통신을 위해서는 O_RDWR mode로 열어야 합니다
      * 서로 다른 프로세스들이 PIPE의 이름만 알면 통신이 가능하다.
      * read()와 write()가 기본적으로 block 모드로 작동하기 때문으로 프로세스가 read 대기중 이라면 read가 끝나기 전에는 write를 할 수가 없게 된다.
         * 참고로 nonblock mode로 열면 대기 안하고 넘어갑니다.

![image](https://github.com/Shortudy-10th/tech-for-developer/assets/32262904/4ddcc6d4-40ca-462c-b37a-4a9a661a6be5)

* Message Queue
   * 정의
      * 메모리를 사용한 PIPE이다.
      * 구조체 기반으로 통신을 한다.
   * 구조
      * FIFO 구조
      * msgtype에 따라 다른 구조체를 가져올 수 있다.
   * 사용시기
      * 프로세스간 다양한 통신을 할 때 사용 할 수 있다.
   * 유의사항
      * 커널에서 제공하는 Message queue 이기 때문에 EnQueue 하는데 제한이 존재 한다.

![image](https://github.com/Shortudy-10th/tech-for-developer/assets/32262904/f4eadfe2-8c62-4bb9-a73a-9ae350e58786)

* Shared Memory
   * 정의
      * 시스템 상의 공유 메모리를 통해 통신한다.
   *  구조
      * 일정한 크기의 메모리를 프로세스간에 공유하는 구조
      * 공유 메모리는 커널에서 관리 된다,
   * 사용시기
      * 프로세스간 Read, Write를 모두 필요로 할 때
   * 유의사항 
      * 프로세스끼리 사용할려면 메모리 크기가 동일 해야 한다.

![image](https://github.com/Shortudy-10th/tech-for-developer/assets/32262904/1ed13c1b-af9b-4a3c-9ab3-3d832cfd5f30)

* Memory Map
   * 정의
      * 파일을 프로세스의 메모리에 일정 부분 맵핑시켜 사용한다.
   * 구조
      * 메모리 맵 파일을 사용하면 디스크에 있는 파일의 블록들과 프로세스의 페이지가 연결된다. 
      * 이 상태에서는 가상 메모리에 사상만 되어 있을 뿐이며, 실제 메모리에 파일을 불러오는 시기는 해당 주소에 읽거나 쓰기 작업을 시도할 때 이루어진다. 
      * 이 작업은 운영 체제에서 가상 메모리를 다루는 데 사용하는 페이징기법과 같이 페이지 부재를 일으키는 방식을 이용한다. 
      * 일반적으로 파일에서 불러오는 자료는 페이지 크기와 같은 4KiB이지만, 몇몇 운영 체제에서는 하드디스크의 낮은 임의 접근 시간을 고려해 4KiB보다 많은 자료를 페이지로 올릴 수도 있다.
   * 사용시기
      * 파일로 대용량 데이터를 공유 할 때 사용한다.
      * FILE IO가 느릴 때 사용하면 좋다.
      * 대부분 운영 체제에서는 프로세스를 실행할 때 실행 파일의 각 세그먼트를 메모리에 사상하기 위해 메모리 맵 파일을 이용한다.
   * 유의 사항
      * write시기는  프로세스의 페이지가 메모리에에서 내려갈 때만 Write된다. 
      * 이렇기 때문에 메모리와 file Sync가 안맞을수 도있다.
      * IA-32 기반 시스템에서 하나의 프로세스에서 PAE 기술을 사용하지 않고 사용 가능한 최대 크기는 4GB로 제한된다.
      * 메모리 맵 파일은 파일의 크기를 바꿀 수는 없으며 메모리 맵 파일을 사용하기 이전, 또는 이후에만 파일의 크기를 바꿀 수 있다.

![image](https://github.com/Shortudy-10th/tech-for-developer/assets/32262904/7fa08641-6825-45df-b05d-833b9469f68c)

* Socket
   * 정의
      * 네트워크 소켓통신을 사용한 데이터 공유
   * 구조
      * 네트워크 소켓을 이용하여 Client - Server 구조로 데이터 통신
   * 사용시기
      * 원격에서 프로세스간 데이터를 공유 할 때 사용
   * 유의 사항
      * 네트워크 프로그래밍이 가능해야 한다.
      * 데이터 세그먼트 처리를 잘해야한다.

![image](https://github.com/Shortudy-10th/tech-for-developer/assets/32262904/ce7f2f5c-c2f5-483a-9223-58a380381ec6)

</blockquote>

<br>

**Reply by sikutisa** [reply]
### ❓️ 스레드 간 통신보다 프로세스 간 통신이 어려운 이유

<br>

### 💬 답변
<blockquote>

스레드는 한 프로세스 내에서 실행되기 때문에 메모리를 공유하는 게 간단하다. 하지만 프로세스는 서로 다른 프로그램이기 때문에 서로 독립적인 메모리 공간을 가지고 있고, 통신을 하기 위해서는 통신을 위한 공간을 따로 만들어주어야 한다. 또한 서로 다른 프로세스이기 때문에 통신할 때 안전을 위해 보안에 신경 써야 한다.

</blockquote>

<br>

**Comment by seongeun42** [comment]
### ❓️ 컨텍스트 스위칭에 대해 설명하고, 프로세스 간에는 어떻게 진행되는지 설명해주세요.

<br>

### 💬 답변
<blockquote>

**컨텍스트 스위칭**
- CPU를 사용 중이던 프로세스가 나가고 새로운 프로세스를 받아들이는 작업으로, 각 프로세스의 PCB 내용이 변경됨

![context_switching](https://github.com/Shortudy-10th/tech-for-developer/assets/64778589/71fe0472-7e02-495c-be84-7084d02a7656)
- P1의 실행 시간이 끝나면 PCB에 현재까지의 작업 결과를 저장하고 P1의 상태는 Ready 혹은 Block 상태가 됨
- Ready 상태였던 P2는 CPU를 할당 받고 CPU에서 P2를 실행함
  - CPU의 레지스터가 P2의 PCB를 가짐


</blockquote>

<br>

**Reply by seongeun42** [reply]
### ❓️ 프로세스의 상태에 대해 설명해주세요.

<br>

### 💬 답변
<blockquote>

 - **프로세스의 5가지 상태**
    <br><img src="https://github.com/Shortudy-10th/tech-for-developer/assets/64778589/6e177bd5-6868-46c1-a189-ac422bce0dbc" width="550" height="350"><br><br>
    - 보류를 포함한 프로세스의 상태
    <br><img src="https://github.com/Shortudy-10th/tech-for-developer/assets/64778589/54806c94-e959-40db-80ec-70b023b8b3ce" width="580" height="350"><br>
    
    |상태||
    |:---:|---|
    |실행(Running)|- CPU를 할당받고 실행 중인 상태|
    |준비(Ready)|- CPU를 할당받기 위해 기다리는 상태 <br> - 실행되기 위한 다른 모든 조건(ex. 메모리 할당)은 만족한 상태로, CPU만 할당받으면 바로 실행 가능한 상태|
    |대기(Blocking)|- 준비가 끝나지 않아서 CPU를 할당받아도 실행될 수 없는 상태 <br> - ex. 프로세스가 요청한 I/O 이벤트가 완료되기를 기다리는 상태 <br> - ex. 공유데이터를 사용하기 위해 기다리는 상태 등|
    |생성(Create)|- 프로세스가 메모리에 올라와 실행 준비를 완료한 상태 <br> - 커널의 주소공간에 PCB가 생성됨|
    |완료(Terminate)|- 작업을 마치고 PCB를 제거한 상태|
    |휴식(Pause)|- 프로세스가 작업을 일시적을 쉬고 있는 상태 <br> - 유닉스에서 프로그램 실행 중 `ctrl` + `z`를 누르면 볼 수 있음|
    |보류(Suspend)|- 프로세스가 메모리에서 잠시 쫓겨난 상태 <br> - 보류 대기 상태와 보류 준비 상태로 나뉨 <br> - 보류 상태와 비교해서 일반적인 프로세스 상태는 활성 상태라고 함 <br> - ex. 메모리가 꽉 차서 일부 프로세스를 메모리 밖으로 내보낼 때 <br> - ex. 프로그램에 오류가 있어 실행을 미룰 때 <br> - ex. 바이러스 등의 악의적인 프로세스로 판단될 때 <br> - ex. 매우 긴 주기로 반복되어 메모리 밖으로 쫓겨나도 괜찮을 때 <br> - ex. 입출력을 기다리는 프로세스의 입출력이 지연될 때|
    
    - 타임 슬라이스(=타임 퀀텀) : 프로세스에 배당된 작업 시간
    - 타임 슬라이스가 다 지나면 클록이 인터럽트를 사용해 CPU에 시간이 다 지났음을 알림
    - CPU 스케줄러 : 준비 상태에 있는 프로세스 중 다음에 실행시킬 프로세스를 선정하는 역할
    - 디스패치(dispatch) : 준비 상태의 프로세스를 실행 상태로 바꾸는 CPU 스케줄러의 작업 
    - 보류 상태에 들어간 프로세스는 메모리 밖으로 쫓겨나 스왑 영역에 보관됨
    - 스왑 영역 : 메모리에서 쫓겨나 데이터가 임시로 보관되는 곳

</blockquote>

<br>

**Comment by seongeun42** [comment]
### ❓️ 프로세스의 구조, 프로세스와 프로그램의 차이점을 말해주세요.

<br>

### 💬 답변
<blockquote>

**프로세스의 구조**
    
![process_memory](https://github.com/Shortudy-10th/tech-for-developer/assets/64778589/6ea6a984-5823-40f0-a49e-98522bb672ce)

|영역||
|:---:|---|
|코드 영역|- 실행 명령을 포함하는 프로그램의 코드들이 저장된 영역 <br> - 읽기 전용|
|데이터 영역|- 코드가 실행되면서 사용하는 변수나 파일 등의 각종 데이터를 모아놓은 영역 <br> - 전역변수와 정적변수가 여기 저장됨 <br> - GVAR : 초기화된 데이터 저장 <br> - BSS : 초기화되지 않은 데이터 저장 <br> - 읽기/쓰기 전용|
|힙 영역|- 필요에 의해 메모리를 동적으로 할당할 때 사용하는 영역|
|스택 영역|- 프로세스를 실행하기 위해 부수적으로 필요한 데이터를 모아놓은 영역 <br> - 함수 호출과 관련된 지역변수와 매개변수 등이 저장됨 <br> - 숨김 영역|
    
- gvar(data)과 bss를 구분하는 이유 : 초기화된 변수는 값도 저장해야 하지만, 초기화 안 된 변수는 메모리 영역만 잡고 값은 저장할 필요가 없어서 구분함
- 힙과 스택은 사실상 같은 공간으로, 힙이 커지면 스택이 작아지고, 힙이 작으면 스택이 커짐


<br>


**프로그램 vs 프로세스**

|프로그램|프로세스|
|:-------:|:-------:|
|저장장치에 저장된 정적인 상태|메모리에 로드된 동적인 상태|
|어떤 데이터를 사용해 어떻게 작업 할지 절차를 적어놓은 것|작성된 작업 절차를 실행에 옮긴 것|
|실행 파일|실행 주체|
    
- 어떤 프로그램이 프로세스로 전환되었다는 것은 운영체제로부터 **PCB**를 받았다는 것을 의미함


</blockquote>

<br>

**Reply by seongeun42** [reply]
### ❓️ 프로세스는 PCB를 가지고 있는 객체입니다. PCB(Process contropl block)은 무엇이고 PCB가 하는 일은 무엇일까요?

<br>

### 💬 답변
<blockquote>

<img src="https://github.com/Shortudy-10th/tech-for-developer/assets/64778589/b6ea367d-efcf-4639-a5a3-c598b6eddfe3" width="200">

**PCB(=TCB : Task Constrol Block)**
- 프로세스를 실행하는 데 필요한 중요한 정보를 보관하는 자료 구조(구조체)
- 모든 프로세스는 고유의 PCB를 가짐
- 프로세스가 생성될 때 PCB가 생성되며, 프로세스가 실행을 완료하면 폐기됨
- PCB는 사용자가 함부로 접근하지 못하도록 보호받는 메모리 영역에 저장됨

</blockquote>

<br>

**Comment by seongeun42** [comment]
### ❓️ 프로세스 종류는 무엇이 있는지 설명해주세요.

<br>

### 💬 답변
<blockquote>

- 운영체제 프로세스
    - 커널 프로세스, 시스템 프로세스라고도 함
    - 모든 시스템 메모리와 프로세서의 명령에 액세스 할 수 있는 프로세스
    - 프로세스 실행 순서 제어
    - 다른 사용자 및 커널 영역을 침범하지 못하게 감시
    - 사용자 프로세스 생성
    
- 사용자 프로세스
    - 사용자 코드를 수행하는 프로세스

</blockquote>

<br>

**Comment by seongeun42** [comment]
### ❓️ 프로세스의 생성 과정에 대해 설명해주세요.

<br>

### 💬 답변
<blockquote>

* Process Creation
   * 프로세스 생성은 부모 프로세스가 연산을 통해 자식 프로세스를 만들어냅니다. 
   * 생성된 자식 프로세스 또한 새로운 자식 프로세스를 만들 수 있으며 이를 구별하기 위해 모든 프로세스는 각자 고유의 PID를 가지게 됩니다. 
   * 이렇게 생성된 프로세스 간의 관계는 하나의 큰 트리구조가 됩니다.
   * init 프로세스 : 리눅스, 유닉스 계열의 OS에서 최초로 실행되는 데몬 프로세스이며 PID는 항상 1입니다.

![image](https://github.com/Shortudy-10th/tech-for-developer/assets/32262904/675bcb7a-9ad1-4bcc-9873-1834ab977604)

* 생성된 자식 프로세스는 각자 고유의 PID, 메모리, CPU 등 새 PCB가 할당되며 고유의 자원을 획득하게 됩니다. 
   * 이로 인하여 부모 프로세스의 자원 접근에 제한이 생기며 특수한 방법을 통해 공유할 수 있게 됩니다.
* 프로세스를 생성한 후 부모 프로세스는 다음과 같이 2가지 행동을 할 수 있습니다.
   * 자식 프로세스가 끝날 때까지 기다립니다. ( -> waiting queue )
   * 자식 프로세스와 함께 동작합니다. (멀티 프로세싱 환경)
   * 자식 프로세스는 다음 중 하나의 프로세스가 됩니다.
* 부모 프로세스와 동일한 새로운 프로세스 : 이 경우 부모 프로세스의 프로그램, 데이터가 완전 복사됩니다.
* 새로운 프로그램 실행 : 새로운 프로그램을 메모리에 load 하고 이를 실행하게 됩니다.

</blockquote>

<br>
